### PR TITLE
wp_preload_links implementation

### PR DESCRIPTION
**Pull request by adamsilverstein**
Previously https://github.com/WordPress/wordpress-develop/pull/2505

Trac ticket: https://core.trac.wordpress.org/ticket/42438
 

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by azaozz** [comment]
LGTM, good job.

File: src/wp-includes/general-template.php [hunk 1]
```diff
@@ -3423,6 +3423,119 @@ function wp_resource_hints() {
...
+
+		foreach ( $unique_resource as $resource_key => $resource_value ) {
+			if ( ! is_scalar( $resource_value ) ) {
+				continue;
+			}
+
+			// Ignore non-supported attributes.
+			if ( ! in_array( $resource_key, array( 'as', 'crossorigin', 'href', 'imagesrcset', 'imagesizes', 'type', 'media' ), true ) &&
+				! is_numeric( $resource_key )
+			) {
```
Comment (azaozz) [inline src/wp-includes/general-template.php]: Just a tiny nitpick but how about:
```
$non_supported_attributes = array( 'as', 'crossorigin', 'href', 'imagesrcset', 'imagesizes', 'type', 'media' );

if ( ! in_array( $resource_key, $non_supported_attributes, true ) && ! is_numeric( $resource_key ) ) { 
  ......
```

File: tests/phpunit/tests/general/wpPreloadLinks.php [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+		$this->assertSame( $expected, $actual );
+	}
+
+	/**
+	 * Test provider for all preload link possible combinations.
+	 *
+	 * @return array[]
+	 */
+	public function provider_preload_resources() {
```
Comment (dream-encode) [inline tests/phpunit/tests/general/wpPreloadLinks.php]:
```suggestion
	public function data_preload_resources() {
```
Use `data` for the prefix per the docs.  (https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/)

File: tests/phpunit/tests/general/wpPreloadLinks.php [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+					),
+				),
+			),
+			'MIME_types'             => array(
+				'expected' => "<link rel='preload' href='https://example.com/style.css' as='style' />\n" .
+							"<link rel='preload' href='https://example.com/video.mp4' as='video' type='video/mp4' />\n" .
+							"<link rel='preload' href='https://example.com/main.js' as='script' />\n",
+				'urls'     => array(
+					array(
+						//Should ignore not valid attributes
```
Comment (dream-encode) [inline tests/phpunit/tests/general/wpPreloadLinks.php]:
```suggestion
						// Should ignore not valid attributes.
```

File: tests/phpunit/tests/general/wpPreloadLinks.php [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+						'href' => 'https://example.com/video.mp4',
+						'as'   => 'video',
+						'type' => 'video/mp4',
+					),
+					array(
+						'href' => 'https://example.com/main.js',
+						'as'   => 'script',
+					),
+					array(
+						//Should ignore not valid attributes.
```
Comment (dream-encode) [inline tests/phpunit/tests/general/wpPreloadLinks.php]:
```suggestion
						// Should ignore not valid attributes.
```

File: tests/phpunit/tests/general/wpPreloadLinks.php [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+						'as'          => 'image',
+						'imagesrcset' => '640.png 640w, 800.png 800w, 1024.png 1024w',
+						'imagesizes'  => '100vw',
+					),
+					array(
+						'href'  => 'https://example.com/image-wide.png',
+						'as'    => 'image',
+						'media' => '(min-width: 601px)',
+					),
+					// No href but not imagesrcset, should be ignored
```
Comment (dream-encode) [inline tests/phpunit/tests/general/wpPreloadLinks.php]:
```suggestion
					// No href but not imagesrcset, should be ignored.
```

File: tests/phpunit/tests/general/wpPreloadLinks.php [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+					array(
+						'as'    => 'image',
+						'media' => '(min-width: 601px)',
+					),
+					// imagesizes is optional
```
Comment (dream-encode) [inline tests/phpunit/tests/general/wpPreloadLinks.php]:
```suggestion
					// imagesizes is optional.
```

File: tests/phpunit/tests/general/wpPreloadLinks.php [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+/**
+ * @group general
+ * @group template
+ * @ticket 42438
+ * @covers ::wp_preload_resources
+ */
+class Tests_General_wpPreloadLinks extends WP_UnitTestCase {
+
+	/**
+	 * @dataProvider provider_preload_resources
```
Comment (dream-encode) [inline tests/phpunit/tests/general/wpPreloadLinks.php]:
```suggestion
	 * @dataProvider data_preload_resources
```

**Review by dream-encode — COMMENTED** [review]
Very minor tweaks.

File: src/wp-includes/general-template.php (hunk 1 above)
Comment (adamsilverstein) [reply src/wp-includes/general-template.php]: 👍🏼 makes for a shorter line and I guess PHP optimizes away the extra variable which was my hesitation. Will update.

**Comment by adamsilverstein** [comment]
> Very minor tweaks.

thanks for the refinements @dream-encode  - applied your suggested changes.

**Comment by manuelRod** [comment]
@adamsilverstein thanks for taking this over, I've been busy and couldn't get into it. LGTM. Happy to see this feature landing to core finally!

**Comment by adamsilverstein** [comment]
Fixed in https://core.trac.wordpress.org/changeset/53846

**Pull request closed without merging**